### PR TITLE
Baselines: multi-label NER for KILE and LIR

### PR DESCRIPTION
**Pull request by uricamic**
Draft of the RoBERTa-base based multi-label NER baseline for both KILE and LIR tasks.

### CODE REVIEW

**Review by uricamic — COMMENTED** [review]
lgtm

File: baselines/NER/docile_pretrain_BERT_onfly.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        # default="unlabeled",
+        metadata={"help": ""},
+    )
+    dataset_id: str = dataclasses.field(
+        default=None,
+        metadata={"help": "The repository id of the dataset to use (via the datasets library)."},
+    )
+    # tokenizer_id: str = dataclasses.field(
+    #     default=None, metadata={"help": "The repository id of the tokenizer to use (via AutoTokenizer)."}
+    # )
```
Comment (uricamic) [inline baselines/NER/docile_pretrain_BERT_onfly.py:54]: Please remove the commented blocks here

File: baselines/NER/docile_pretrain_BERT_onfly.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    output_dir: str = dataclasses.field(
+        default=Path("/storage/table_extraction/pretrain_roberta/"),
+        metadata={
+            "help": "The repository id where the model will be saved or loaded from for further pre-training."
+        },
+    )
+    # hf_hub_token: str = dataclasses.field(
+    #     default=False,
+    #     metadata={"help": "The Token used to push models, metrics and logs to the Hub."},
+    # )
```
Comment (uricamic) [inline baselines/NER/docile_pretrain_BERT_onfly.py:64]: and here

File: baselines/NER/docile_pretrain_BERT_onfly.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    # set seed for reproducibility
+    seed = 34
+    set_seed(seed)
+
+    # # load processed dataset
+    # # train_dataset = load_dataset(script_args.dataset_id, split="train")
+    # train_dm = DataMaker()
+    # train_dataset = train_dm.as_hf_dataset(tokenizer)
+    # load trained tokenizer
+    # tokenizer = AutoTokenizer.from_pretrained(script_args.tokenizer_id)
```
Comment (uricamic) [inline baselines/NER/docile_pretrain_BERT_onfly.py:261]: These can be removed now

File: baselines/NER/docile_train_NER_multilabel.py [hunk 4]
```diff
@@ -0,0 +1,1165 @@
...
+            #     flag = True
+            #     first_row = 0
+            #     for row in table["rows"]:
+            #         sep = row["top_position"]*sfy
+            #         if flag:
+            #             first_row = sep
+            #             flag = False
+            #         row_sep[-1].append(sep)
+            #     sep = first_row + (t_h*sfy)
+            #     row_sep[-1].append(sep)
```
Comment (uricamic) [inline baselines/NER/docile_train_NER_multilabel.py:495]: Can be removed now

File: baselines/NER/my_roberta_multilabel.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+            r = bboxes[:, :, 2]
+            b = bboxes[:, :, 3]
+            cx = (l + r) / 2
+            cy = (t + b) / 2
+            w = r - l
+            h = b - t
+            pos2_cx_emb = self.pos2_cx_emb((cx / self.quant_step_size + 0.5).int())
+            pos2_cy_emb = self.pos2_cx_emb((cy / self.quant_step_size + 0.5).int())
+            pos2_w_emb = self.pos2_cx_emb((w / self.quant_step_size + 0.5).int())
+            pos2_h_emb = self.pos2_cx_emb((h / self.quant_step_size + 0.5).int())
```
Comment (uricamic) [inline baselines/NER/my_roberta_multilabel.py:377]: There is a bug in here, `pos2_cx_emb` is used everywhere instead of using the correct `pos2_X_emb`, respectively

**Review by uricamic — COMMENTED** [review]
Just some minor comments, otherwise LGTM

File: baselines/NER/README.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# NER baselines
+
+Our baselines build upon Named Entity Recognition (NER). Here we provide the code to construct the targets for NER from the DocILE dataset, train RoBERTa and LayoutLMv3 models (including RoBERTa pretraining on the unlabeled part of dataset) and run inference, combining the model predictions for individual OCR tokens into the predicted fields. See [dataset and benchmark paper](../../README.md#dataset-and-benchmark-paper) for more details.
```
[the remark this answers is not in the document]
Comment (simsa-st) [reply baselines/NER/README.md:3]: yes but it will be in a different folder as mentioned in the parent README file

File: baselines/NER/docile_inference_NER_multilabel.py [hunk 7]
```diff
@@ -0,0 +1,968 @@
...
+    "line_item_person_name",
+    "line_item_position",
+    "line_item_quantity",
+    "line_item_tax",
+    "line_item_tax_rate",
+    "line_item_unit_price_gross",
+    "line_item_unit_price_net",
+    "line_item_units_of_measure",
+    "line_item_weight",
+]
```
[the remark this answers is not in the document]
Comment (simsa-st) [reply baselines/NER/docile_inference_NER_multilabel.py:84]: You are right, I am happy to do that, there will be just one tiny difference, KILE_CLASSES will have a slightly different order (due to `payment_reference` not being in alphabetical order in your case as it replaced `variable_symbol`). Could that somehow affect results/trainings?

File: baselines/NER/docile_inference_NER_multilabel.py [hunk 8]
```diff
@@ -0,0 +1,968 @@
...
+    parser.add_argument(
+        "--checkpoint",
+        type=str,
+    )
+    parser.add_argument(
+        "--output_dir",
+        # default=Path("/mnt/shared/ailabs/table_extraction/predictions/NER/from_table_crops/docile221221-0/XLM_RoBERTa_Large"),
+        default=Path(
+            "/mnt/shared/ailabs/table_extraction/predictions/NER/from_table_crops/docile221221-0/RoBERTa_base"
+        ),
```
[the remark this answers is not in the document]
Comment (simsa-st) [reply baselines/NER/docile_inference_NER_multilabel.py:414]: I will remove them or replace them with the `/app/data/...` defaults

File: baselines/NER/docile_inference_NER_multilabel.py [hunk 9]
```diff
@@ -0,0 +1,968 @@
...
+        args.checkpoint, config=config
+    ).to(device)
+
+    # fix the old models (variable_symbol -> payment_reference)
+    try:
+        print("INFO: model with an obsolete label set. Updating the label-set...")
+        model.config.id2label[model.config.label2id["B-variable_symbol"]] = "B-payment_reference"
+        model.config.id2label[model.config.label2id["I-variable_symbol"]] = "I-payment_reference"
+    except Exception:
+        print("INFO: model with up-to-date label set")
```
[the remark this answers is not in the document]
Comment (simsa-st) [reply baselines/NER/docile_inference_NER_multilabel.py:477]: Unfortunately it is still needed, the roberta base checkpoint we use still has variable symbol in the config. We can get rid of this later if we retrain all baselines with the unified parameters.

File: baselines/NER/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (uricamic) [reply baselines/NER/README.md:3]: Ok 😄

File: baselines/NER/docile_inference_NER_multilabel.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (uricamic) [reply baselines/NER/docile_inference_NER_multilabel.py:477]: I see, ok then 👍

File: baselines/NER/docile_inference_NER_multilabel.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (uricamic) [reply baselines/NER/docile_inference_NER_multilabel.py:84]: It should not have any impact on the trained models, because the classes should be alphabetically ordered inside the model (see e.g. https://github.com/rossumai/docile/blob/fb91d582b9466f8fc5f3cedd922e160856390938/baselines/NER/docile_train_NER_multilabel.py#L857)

File: pyproject.toml [hunk 10]
```diff
@@ -1,11 +1,12 @@
...
 name = "docile-benchmark"
-version = "0.2.2"
+version = "0.3.0"
 description = "Tools to work with the DocILE dataset and benchmark"
 authors = [
     "Stepan Simsa <stepan.simsa@rossum.ai>",
     "Matyas Skalicky <matyas.skalicky@rossum.ai>",
-    "Milan Sulc <milan.sulc@rossum.ai>"
+    "Milan Sulc <milan.sulc@rossum.ai>",
+    "Michal Uricar <michal.uricar@rossum.ai>",
```
[the remark this answers is not in the document]
Comment (simsa-st) [reply pyproject.toml:9]: I thought Matej and Yash will add themselves in their PRs but I guess it will be easier to add both of them here 👍

File: baselines/NER/docile_inference_NER_multilabel.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (simsa-st) [reply baselines/NER/docile_inference_NER_multilabel.py:84]: I see, great! It also definitely should not impact inference (there the order is not used at all).

**Comment by uricamic** [comment]
LGTM

**Review by simsa-st — APPROVED** [review]
